akka-actor-typed/src/main/scala/akka/actor/typed/internal/receptionist/LocalReceptionist.scala

```diff
@@ -37,98 +38,200 @@ private[akka] object LocalReceptionist extends ReceptionistBehaviorProvider {
+  private type Service[K <: AbstractServiceKey] = ActorRef[K#Protocol]
+  private type Subscriber[K <: AbstractServiceKey] = ActorRef[ReceptionistMessages.Listing[K#Protocol]]
 
-  sealed trait InternalCommand
-  final case class RegisteredActorTerminated[T](key: ServiceKey[T], ref: ActorRef[T]) extends InternalCommand
-  final case class SubscriberTerminated[T](key: ServiceKey[T], ref: ActorRef[ReceptionistMessages.Listing[T]])
+  private sealed trait InternalCommand
+  private final case class RegisteredActorTerminated[T](ref: ActorRef[T]) extends InternalCommand
+  private final case class SubscriberTerminated[T](ref: ActorRef[ReceptionistMessages.Listing[T]])
       extends InternalCommand
 
-  override def behavior: Behavior[Command] = Behaviors.setup { ctx =>
-    ctx.setLoggerName(classOf[LocalReceptionist])
-    behavior(TypedMultiMap.empty[AbstractServiceKey, KV], TypedMultiMap.empty[AbstractServiceKey, SubscriptionsKV])
-      .narrow[Command]
+  private object State {
+    def empty =
+      State(
+        TypedMultiMap.empty[AbstractServiceKey, Service],
+        Map.empty,
+        TypedMultiMap.empty[AbstractServiceKey, Subscriber],
+        Map.empty)
   }
+  private final case class State(
+      services: TypedMultiMap[AbstractServiceKey, Service],
+      servicesPerActor: Map[ActorRef[_], Set[AbstractServiceKey]], // needed since an actor can implement several services
+      subscriptions: TypedMultiMap[AbstractServiceKey, Subscriber],
+      subscriptionsPerActor: Map[ActorRef[_], Set[AbstractServiceKey]] // needed since a subscriber can subscribe to several keys
+  ) {
+
+    def serviceInstanceAdded[Key <: AbstractServiceKey, SI <: Service[Key]](key: Key)(serviceInstance: SI): State = {
+      val newServices = services.inserted(key)(serviceInstance)
+      val newServicePerActor =
+        servicesPerActor.updated(
+          serviceInstance,
+          servicesPerActor.getOrElse(serviceInstance, Set.empty) + key.asServiceKey)
+      copy(services = newServices, servicesPerActor = newServicePerActor)
+    }
+
+    def serviceInstanceRemoved[Key <: AbstractServiceKey, SI <: Service[Key]](key: Key)(serviceInstance: SI): State = {
+      val newServices = services.removed(key)(serviceInstance)
+      val newServicePerActor =
+        servicesPerActor.get(serviceInstance) match {
+          case Some(keys) =>
+            val newKeys = keys - key.asServiceKey
+            // only/last service this actor was registered for
+            if (newKeys.isEmpty) {
+              servicesPerActor - serviceInstance
+            } else servicesPerActor.updated(serviceInstance, newKeys)
+          case None =>
+            // no services actually registered for actor
+            servicesPerActor
+        }
+      copy(services = newServices, servicesPerActor = newServicePerActor)
+    }
+
+    def serviceInstanceRemoved(serviceInstance: ActorRef[_]): State = {
```

Comment:
Will both removed functions - by key or instance - be needed?

Reply:
Yes, because an service providing actor could register itself for several service keys, and then unregister for one of them, or die, which means it should be removed for all keys it has registered for.

Reply:
Yes, because an service providing actor could register itself for several service keys, and then unregister for one of them, or die, which means it should be removed for all keys it has registered for.